Rewrite section about custom context flavors (#388)
grammyjs/website: site/docs/guide/context.md

> If you are new to context objects, you don't need to worry about the rest of this page.

You can install your own properties on the context object if you want.

### Via Middleware (Recommended)

> The customizations can easily be done in [middleware](./middleware.md).

Reviewer: ```suggestion
The customizations can be easily done in [middleware](./middleware.md).
```

::: tip Middlewhat?

> This section requires an understanding of middleware, so in case you did not skip ahead to [this section](./middleware.md) yet, here is a very brief summary.

Reviewer: ```suggestion
This section requires an understanding of middleware, so in case you have not skipped ahead to [this section](./middleware.md) yet, here is a very brief summary.
```

All you really need to know is that several handlers can process the same context object.
There are special handlers which can modify `ctx` before any other handlers are run, and the modifications of the first handler will be visible to all subsequent handlers.
:::

The idea is to install middleware before you register other listeners.
You can then set the properties you want inside these handlers.

> For illustration purposes, let's say you want to set a property `ctx.config` on the context object.

Reviewer: ```suggestion
For illustration purposes, let's say you want to set a property called `ctx.config` on the context object.
```

Right after creating your bot, do this:

```ts
const BOT_DEVELOPER = 123456; // bot developer chat identifier

bot.use(async (ctx, next) => {
  // Modify context object here by setting the config.
  ctx.config = {
    developer: BOT_DEVELOPER,
    isDeveloper: ctx.from?.id === BOT_DEVELOPER,
  };
  // Run remaining handlers.
  await next();
});
```

After that, you can use `ctx.config` in the remaining handlers.

```ts
bot.command("start", async (ctx) => {
  // Work with modified context here!
  if (ctx.config.isDeveloper) await ctx.reply("Hi mom!! <3");
  else await ctx.reply("Welcome, human!");
});
```

However, you will notice that TypeScript does not know that `ctx.config` is available, even though we are assigning the property correctly.
So while the code will work at runtime, it does not compile.
To fix this, we need to adjust the type of the context and add the property.

```ts
interface BotConfig {
  botDeveloper: number;
  isDeveloper: boolean;
}

type MyContext = Context & {
  config: BotConfig;
};
```

The new type `MyContext` now accurately describes the context objects our bot is actually handling.

> > You need to make sure that you keep the types in sync with the properties you initialize.

Reviewer: ```suggestion
> You will need to make sure that you keep the types in sync with the properties you initialize.
```
